docs/architecture/grpc-for-wcf-developers/migrating-duplex-services.md

```diff
@@ -1,11 +1,481 @@
+
+    try { await actionsTask; } catch { /* Ignored */ }
+
+    _logger.LogInformation("Subscription finished.");
+}
+
+private async Task WriteUpdateAsync(IServerStreamWriter<StockTickerUpdate> stream, string symbol, decimal price)
+{
+    try
+    {
+        await stream.WriteAsync(new StockTickerUpdate
+        {
+            Symbol = symbol,
+            PriceCents = (int)(price * 100),
+            Time = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow)
+        });
+    }
+    catch (Exception e)
+    {
+        // Handle any errors due to broken connection etc.
+        _logger.LogError($"Failed to write message: {e.Message}");
+    }
+}
+
+private static Task AwaitCancellation(CancellationToken token)
+{
+    var completion = new TaskCompletionSource<object>();
+    token.Register(() => completion.SetResult(null));
+    return completion.Task;
+}
+```
+
+The `ActionMessage` class that gRPC has generated for us guarantees that only one of the `Add` and `Remove` properties can be set, and finding which one is not `null` is a valid way of finding which type of message is used, but there is a better way. The code generation also created an `enum ActionOneOfCase` in the `ActionMessage` class, which looks like this:
+
+```csharp
+public enum ActionOneofCase {
+    None = 0,
+    Add = 1,
+    Remove = 2,
+}
+```
+
+The property `ActionCase` on the `ActionMessage` object can be used with a `switch` statement to determine which field is set.
+
+```csharp
+private async Task HandleActions(IAsyncStreamReader<ActionMessage> requestStream, IFullStockPriceSubscriber subscriber, CancellationToken token)
+{
+    await foreach (var action in requestStream.ReadAllAsync(token))
+    {
+        switch (action.ActionCase)
+        {
+            case ActionMessage.ActionOneofCase.None:
+                _logger.LogWarning("No Action specified.");
+                break;
+            case ActionMessage.ActionOneofCase.Add:
+                subscriber.Add(action.Add.Symbol);
+                break;
+            case ActionMessage.ActionOneofCase.Remove:
+                subscriber.Remove(action.Remove.Symbol);
+                break;
+            default:
+                _logger.LogWarning($"Unknown Action '{action.ActionCase}'.");
+                break;
+        }
+    }
+}
+```
+
+> [!TIP]
+> The `switch` statement has a `default` case that logs a warning if an unknown `ActionOneOfCase` value is encountered. This could be useful in indicating that a client is using a later version of the `.proto` file which has added more actions. This is one reason why using a `switch` is better than testing for `null` on known fields.
+
+### Using the FullStockTickerService from a client application
+
+There is a simple .NET Core 3.0 WPF application to demonstrate use of this more complex client. The full application can be found [on GitHub](https://github.com/RendleLabs/grpc-for-wcf-developers/tree/master/FullStockTickerSample/grpc/FullStockTicker).
```

Comment:
is this a private repo?

Reply:
Yeah, I'm not sure where to put all the code so it's accessible to readers. Will you host it in a Microsoft org?